src/avro_copy.cpp

```diff
@@ -0,0 +1,498 @@
+		return str.GetSize();
+	}
+	case LogicalTypeId::ENUM: {
+		//! TODO: add support for ENUM
+		throw NotImplementedException("Can't convert ENUM Value (%s) to Avro yet", val.ToString());
+	}
+	case LogicalTypeId::MAP:
+	case LogicalTypeId::LIST: {
+		auto &list_values = ListValue::GetChildren(val);
+		idx_t list_value_size = 0;
+		for (idx_t i = 0; i < list_values.size(); i++) {
+			auto &list_value = list_values[i];
+
+			avro_value_t item;
+			size_t unused_new_index;
+			if (avro_value_append(target, &item, &unused_new_index)) {
+				throw InvalidInputException(avro_strerror());
+			}
+
+			list_value_size += PopulateValue(&item, list_value);
+		}
+		return list_value_size + 1;
+	}
+	case LogicalTypeId::STRUCT: {
+		auto &struct_values = StructValue::GetChildren(val);
+		idx_t struct_value_size = 0;
+		for (idx_t i = 0; i < struct_values.size(); i++) {
+			const char *unused_name;
+			avro_value_t field;
+			if (avro_value_get_by_index(target, i, &field, &unused_name)) {
+				throw InvalidInputException(avro_strerror());
+			}
+			struct_value_size += PopulateValue(&field, struct_values[i]);
+		}
+		return struct_value_size + 1;
+	}
+	default:
+		throw NotImplementedException("PopulateValue not implemented for type %s", type.ToString());
+	}
+}
+
+static void WriteAvroSink(ExecutionContext &context, FunctionData &bind_data_p, GlobalFunctionData &gstate_p,
+                          LocalFunctionData &lstate_p, DataChunk &input) {
+	auto &global_state = gstate_p.Cast<WriteAvroGlobalState>();
+	auto &local_state = lstate_p.Cast<WriteAvroLocalState>();
+
+	auto &datum_buffer = global_state.datum_buffer;
+	idx_t count = input.size();
+	idx_t offset_in_datum_buffer = 0;
+	for (idx_t i = 0; i < count; i++) {
+
+		//! Populate our avro value, estimating the size of the value as we go
+		idx_t value_size = 0;
+		for (idx_t col_idx = 0; col_idx < input.ColumnCount(); col_idx++) {
+			auto val = input.GetValue(col_idx, i);
+
+			const char *unused_name;
+			avro_value_t column;
+			if (avro_value_get_by_index(&local_state.value, col_idx, &column, &unused_name)) {
+				throw InvalidInputException(avro_strerror());
+			}
+			value_size += PopulateValue(&column, val);
+		}
+
+		//! Prepare the datum buffer for this row
+		idx_t length = datum_buffer.GetCapacity() - offset_in_datum_buffer;
+		if (value_size > length) {
+			//! This value is too big to fit into the remaining portion of the buffer
+			idx_t new_capacity = datum_buffer.GetCapacity();
+			new_capacity += value_size;
+			datum_buffer.ResizeAndCopy(NextPowerOfTwo(new_capacity));
+		}
+		avro_writer_memory_set_dest_with_offset(global_state.datum_writer, (const char *)datum_buffer.GetData(),
+		                                        datum_buffer.GetCapacity(), offset_in_datum_buffer);
+
+		while (avro_file_writer_append_value(global_state.file_writer, &local_state.value)) {
+			auto current_capacity = datum_buffer.GetCapacity();
```

Comment:
nit: could we maybe check the error here? I would not be surprised if an error exists where doubling the buffer is not the solution, and throwing some other exception is better